### PR TITLE
[java] Extract reaching definitions analysis out of UnusedAssignmentRule

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

This way the data-flow information it collects can be used by more rules. This reimplements several rules using this information.
SingularField and ImmutableField should be more precise and more general. MissingBreakInSwitch is updated according to #2894, ie, its implementation strategy uses control flow information to detect fall-through, instead of blindly counting breaks. The remaining piece to fix this issue, is to rename the rule to "FallThroughSwitchCase".

Also part of these changes aim to reduce the importance of AbstractLombokAwareRule, by identifying the actual annotations that each rule should care about instead of just ignoring all lombok annotations for all those rules.

## Related issues

- Relates to #2701 

Updates rules:
* AvoidThrowingNullPointerException
* ImmutableField
* SingularField
* MissingBreakInSwitch

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 1054 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/68176b603a68471389b7ab394059fe56/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 9674 violations,<br />
introduces 12040 new violations, 1 new errors and 0 new configuration errors,<br />
removes 16174 violations, 8 errors and 2 configuration errors.<br />
<a href="https://chunk.io/pmd/68176b603a68471389b7ab394059fe56/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/MissingBreakInSwitch.xml [hunk 1]
```diff
@@ -217,6 +217,63 @@ public class DemoMissingBreakContinue {
...
+            // fallthrough
+            default: break;
+            }
+        }
+    }
+}
+        ]]></code>
+    </test-code>
+    <test-code>
+        <description>Suppresswarnings</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/MissingBreakInSwitch.xml:262]:
```suggestion
        <description>[java] Recognize @SuppressWanings("fallthrough") for MissingBreakInSwitch #1899</description>
```

**Review by adangel — APPROVED** [review]
Thanks!

At some point, we should create unit tests for the dataflow implementation....

As far as I see, this PR

- fixes #1899 

I'll add this to the release notes.

The other referenced missingbreakinswitch bugs are still open, I assume (at least, there are no test cases):
- #659 still open
- #2579 there is no test, but in github project, it is marked as done?

Then there is the renaming of "MissingBreakInSwitch" into "ImplicitSwitchFallThrough" (Problem 2 of #2894).
I'd suggest to make this a separate issue. Then we can rename it already in PMD 6 and remove the deprecated
rule reference in PMD 7.

**Comment by adangel** [comment]
> The other referenced missingbreakinswitch bugs are still open, I assume (at least, there are no test cases):
> -  #659 still open
> -  #2579 there is no test, but in github project, it is marked as done?

Ok, I tested them. You've fixed them already.

**Comment by oowekyala** [comment]
Thanks for the review. I'll check whether the bugs of MissingBreakInSwitch are already solved and add tests if so

**Pull request merged**